compiler/zrc_codegen/src/program.rs

```diff
@@ -26,6 +26,65 @@ use crate::{
     ty::{create_fn, llvm_basic_type, llvm_type},
 };
 
+/// Evaluate a constant expression to an LLVM constant value.
+/// This is used for global variable initializers.
+///
+/// # Panics
+/// Panics if the expression is not a valid constant expression.
+#[allow(clippy::too_many_lines, clippy::wildcard_enum_match_arm)]
+fn eval_const_expr<'ctx>(
+    unit: &CompilationUnitCtx<'ctx, '_>,
+    expr: &zrc_typeck::tast::expr::TypedExpr,
+    ty: &Type,
+) -> BasicValueEnum<'ctx> {
+    use inkwell::types::StringRadix;
+    use zrc_typeck::tast::expr::{NumberLiteral, TypedExprKind};
+
+    match expr.kind.value() {
+        TypedExprKind::NumberLiteral(n, _) => {
+            let no_underscores = match n {
+                NumberLiteral::Decimal(string)
+                | NumberLiteral::Binary(string)
+                | NumberLiteral::Hexadecimal(string) => string.replace('_', ""),
+            };
+
+            let radix = match n {
+                NumberLiteral::Decimal(_) => StringRadix::Decimal,
+                NumberLiteral::Binary(_) => StringRadix::Binary,
+                NumberLiteral::Hexadecimal(_) => StringRadix::Hexadecimal,
+            };
+
+            let (llvm_ty, _) = llvm_basic_type(unit, ty);
+            llvm_ty
+                .into_int_type()
+                .const_int_from_string(&no_underscores, radix)
+                .expect("number literal should have parsed correctly")
+                .as_basic_value_enum()
+        }
+        TypedExprKind::BooleanLiteral(value) => unit
+            .ctx
+            .bool_type()
+            .const_int((*value).into(), false)
+            .as_basic_value_enum(),
+        TypedExprKind::StringLiteral(string) => {
+            let bytes = string.as_bytes();
+            unit.ctx
+                .const_string(bytes.as_bytes(), false)
+                .as_basic_value_enum()
+        }
+        TypedExprKind::CharLiteral(ch) => unit
+            .ctx
+            .i8_type()
+            .const_int(ch.as_byte().into(), false)
+            .as_basic_value_enum(),
+        _ => {
+            // For now, unsupported constant expressions default to zero
```

Comment:
This produces behavior inconsistent with Zirco's design goals. This should be a type check error.